### ISSUE TITLE
uv self update: failed to execute installer (status: exit code: 1) on Windows

### ISSUE THREAD

**Issue by tschneiderr** [astral-sh/uv#5460]
When running `uv self update --verbose` on my Windows 11 machine, I get the following output:

```
DEBUG uv 0.2.28
info: Checking for updates...
error: failed to execute installer (status: exit code: 1)
```

Installing and updating `uv` through the following command works as expected: `irm https://astral.sh/uv/install.ps1 | iex`
`uv` has only been installed via the standalone installer, not via `pip`.

I can't seem to reproduce the issue on any other system or virtual machine.

**Comment by zanieb** [comment]
Thanks for the report!

**Comment by zanieb** [comment]
What shell are you using?

**Comment by tschneiderr** [comment]
I'm using the current PowerShell v7.4.4.
But switching to the built-in PowerShell v5 from Windows produces the same result.

**Comment by zanieb** [comment]
Unfortunately I'm not sure what's going on. I'm chatting with maintainers of the upstream dependency (`cargo-dist`) which creates the installer for us.

Does it persist if you delete and reinstall?

**Comment by charliermarsh** [comment]
Do you mind testing with the latest version? It should have a change that surfaces more information in the trace.

**Comment by FishAlchemist** [comment]
@tschneiderr I couldn't reproduce your issue on my Windows 11 system either. So I'm wondering if it could be an instruction set architecture issue? Windows 11 currently comes in both ARM64 and x86-64 versions. My Windows 11 is the x86-64 version.

**Comment by tschneiderr** [comment]
I updated `uv` to v0.2.30 via the following command, which worked as before:

```
irm https://astral.sh/uv/0.2.30/install.ps1 | iex
```

Running `uv self update --verbose` now doesn't give any additional information:

```
DEBUG uv 0.2.30
info: Checking for updates...
error: failed to execute installer (status: exit code: 1)
```

@charliermarsh Is there any way for me to provide more detailed logs of what's going wrong?
@FishAlchemist I'm running Windows on x86-64.

**Comment by tschneiderr** [comment]
Trying to update `uv` `v0.2.36` via `uv self update --verbose` revealed some information which helped me fix the problem.

```pwsh
DEBUG uv 0.2.36
info: Checking for updates...
error: The installation failed. Output from the installer:
C:\Users\Thomas\AppData\Local\Temp\.tmpCelC4m\installer.ps1 : File
C:\Users\Thomas\AppData\Local\Temp\.tmpCelC4m\installer.ps1 cannot be loaded because running scripts is disabled on
this system. For more information, see about_Execution_Policies at https:/go.microsoft.com/fwlink/?LinkID=135170.
At line:1 char:1
+ C:\Users\Thomas\AppData\Local\Temp\.tmpCelC4m\installer.ps1
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : SecurityError: (:) [], PSSecurityException
    + FullyQualifiedErrorId : UnauthorizedAccess
```

My `Execution-Policy` is set to `RemoteSigned`. The `uv` standalone installer also works, so what's going on here?

`Get-ExecutionPolicy -List`

```pwsh
        Scope ExecutionPolicy
        ----- ---------------
MachinePolicy       Undefined
   UserPolicy       Undefined
      Process       Undefined
  CurrentUser    RemoteSigned
 LocalMachine    RemoteSigned
```

It turnes out that the current `PowerShell Core (v7)` and the integrated `PowerShell Classic (v5)` have separate `Execution-Policy` settings and for some reason the integrated `uv` updater runs in `PowerShell Classic`!

`powershell.exe Get-ExecutionPolicy -List`

```pwsh
        Scope ExecutionPolicy
        ----- ---------------
MachinePolicy       Undefined
   UserPolicy       Undefined
      Process       Undefined
  CurrentUser       Undefined
 LocalMachine       Undefined
```

Changing the `Execution-Policy` to `RemoteSigned` in `PowerShell Classic` too fixes my issues! 😄

I think the updater should run in the same shell as the `uv` command is run from to prevent there kind of issues.

**Comment by zanieb** [comment]
Thanks for the details! The updater itself is managed outside of uv — we'll raise this with the Axo team.

**Comment by jon-ton** [comment]
I had a similar problem with Powershell on windows. What worked for me was the following command to avoid spawning the shell within a shell:

`Invoke-WebRequest https://astral.sh/uv/install.ps1 -UseBasicParsing | iex`

**Comment by zanieb** [comment]
@jon-ton would you be willing to contribute that to the installation docs?

**Comment by jon-ton** [comment]
Ah, sorry, I'm a reluctant Windows user. 

It ended up being even easier. If using PowerShell core:
`irm https://astral.sh/uv/install.ps1 | iex`

Whereas the original uses the old windows PowerShell:
  `powershell -c "irm https://astral.sh/uv/install.ps1 | iex"`

Is it safe to assume people should be using the new PowerShell core and just replace the installation instructions, or list both?

**Comment by zanieb** [comment]
I'm not sure — these recommendations are from an external dependency which creates these installers for us. I'll check in there.

**Comment by Gankra** [comment]
I agree it would be great to just run in the current version of powershell, core on not. Unfortunately, explicitly invoking powershell here is a workaround for the reverse problem: people would paste this command into cmd and it would fail. And microsoft quite explicitly made classic powershell and powershell core have different commands (`powershell` vs `pwsh`) to make sure you don't accidentally invoke one when you meant the other, since they have fairly different semantics (especially around utf8 vs utf16).

I'm pretty confident the cmd case is more pervasive than the pwsh one, but still, it would be good for us to find a way to manage this execution policy thing better.

**Comment by 2-5** [comment]
I've hit this too. I have a pretty default Windows 11, with PowerShell core, but with dev mode disabled:

```powershell
PowerShell 7.4.4

PS C:\Users\user> powershell -c "irm https://astral.sh/uv/install.ps1 | iex"
Error: PowerShell requires an execution policy in [Unrestricted, RemoteSigned, ByPass]
to run uv. For example, to set the execution policy to 'RemoteSigned' please run:

    Set-ExecutionPolicy RemoteSigned -scope CurrentUser

PS C:\Users\user> irm https://astral.sh/uv/install.ps1 | iex
Downloading uv 0.3.3 (x86_64-pc-windows-msvc)
Installing to C:\Users\user\.cargo\bin
  uv.exe
  uvx.exe
everything's installed!

C:\Users\user\.cargo\bin was added to your PATH, you may
need to restart your shell for that to take effect.
```

The alternative ways of installing on Windows should at least be documented:

```powershell
powershell -c "irm https://astral.sh/uv/install.ps1 | iex"
powershell -ExecutionPolicy Bypass -c "irm https://astral.sh/uv/install.ps1 | iex"
irm https://astral.sh/uv/install.ps1 | iex
```

It might be worth mentioning that under `cmd` one should use the `powershell -c` variant, and the `irm ...` variant for `powershell`.

Windows also ships with `curl` now, so a variation of this would work under both `cmd` and `powershell`. The `.bat` file would autodetect the env it's run under and proceed accordingly:

```shell
curl -LsSf https://astral.sh/uv/install.bat -o uv_install.bat && .\uv_install.bat
```

**Comment by FishAlchemist** [comment]
curl has different behaviors in windows. It is recommended to add **.exe** to explicitly use ``curl.exe``.
* Powershell 5.1 (Windows Powershell) 
![image](https://github.com/user-attachments/assets/2dd4444a-ebbf-4b30-9040-4cfb033545cd)
* Powershell 5.1 (Windows Powershell) with .exe
![image](https://github.com/user-attachments/assets/46bc874b-1546-4c97-9483-a7c2323bd287)
* PowerShell 7 (PowerShell Core)
![image](https://github.com/user-attachments/assets/b1613253-ae56-41c5-a471-77270fe7bd0b)
* Command Prompt (cmd.exe)
![image](https://github.com/user-attachments/assets/39b22dbb-5c8c-4fe5-81ba-06fc780fee90)

**Link:**
[Invoke-WebRequest](https://learn.microsoft.com/en-us/powershell/module/microsoft.powershell.utility/invoke-webrequest) (Microsoft Learn)
[curl shipped by Microsoft](https://curl.se/windows/microsoft.html) (curl website)

**Comment by kylewardlow** [comment]
There is a similar error for users for whom the Powershell ExecutionPolicy is dictated by corporate client security group policies and thus inescapable. A workaround I've found is to simply use git bash and `curl -LsSf https://astral.sh/uv/install.sh | sh`, but a functional Powershell method that works for all users, especially those in security-controlled environments, should be implemented.

**Comment by tkmeyer** [comment]
I once had the bad idea to manually uninstall powershell 5.1, so I was only left with pwsh and that left me with the same bug this isssue was initially opened for. So I strongly support the idea to let  the updater run in the same shell as the uv command is run from and to also mention 

`pwsh -ExecutionPolicy Bypass -c "irm https://astral.sh/uv/install.ps1 | iex"`

as another alternative way to install uv in the docs.
